vpr/src/base/xdc_constraints.cpp

```diff
@@ -0,0 +1,244 @@
+        
+        auto* port = tcl_obj_getptr<AtomPortId>(tcl_port);
+        if (port == nullptr)
+            return this->_ret_error("set_property: port__name of PACKAGE_PIN should have a `port` type.");
+
+        return this->_do_set_property_package_pin(pin_name, *port);
+    }
+
+    int _do_set_property_package_pin(const char* pin_name, AtomPortId port) {
+        /* Find associated AtomBlock */
+
+        const auto& phys_grid_mapping = this->arch.phys_grid_mapping;
+        
+        std::string pin_str(pin_name);
+        auto it = phys_grid_mapping.find(pin_str);
+        if (it == phys_grid_mapping.end())
+            return this->_ret_error("Can't find physical PIN named `" +
+                                    pin_str + "`");
+        const t_phys_map_region& package_pin_region = it->second;
+        
+        AtomBlockId port_block = this->netlist.port_block(port);
+
+        /* Create a 1x1 VprConstraints Partition to contrain the block */
+
+        Partition part;
+        part.set_name(pin_str + ".PART");
+        Region r;
+        r.set_region_rect(
+            package_pin_region.x,
+            package_pin_region.y,
+            package_pin_region.x + package_pin_region.w - 1,
+            package_pin_region.y + package_pin_region.h - 1
+        );
+        r.set_sub_tile(package_pin_region.subtile);
+        PartitionRegion pr;
+        pr.add_to_part_region(r);
+        part.set_part_region(pr);
+        PartitionId part_id = this->constraints.add_partition(part);
+
+        /* Constrain the AtomBlock */
+
+        this->constraints.add_constrained_atom(port_block, part_id);
+
+        VTR_LOG("XDC: Assigned PACKAGE_PIN %s = [%d, %d, %d, %d].%d\n",
+                pin_name, package_pin_region.x, package_pin_region.y,
+                package_pin_region.w, package_pin_region.h, package_pin_region.subtile);
+
+        return this->_ret_ok();
+    }
+
+    int _set_property_iostandard(Tcl_Obj* tcl_io_standard, Tcl_Obj* tcl_port) {
+        const char* io_standard = Tcl_GetString(tcl_io_standard);
+        if (tcl_io_standard == nullptr)
+            return this->_ret_error("set_property: iostandard should be a string.");
+        
+        auto* port = tcl_obj_getptr<AtomPortId>(tcl_port);
+        if (port == nullptr)
+            return this->_ret_error("set_property: port_name of IOSTANDARD should have a `port` type.");
+        
+        return this->_do_set_property_iostandard(io_standard, *port);
+    }
+
+    int _do_set_property_iostandard(const char* io_standard, AtomPortId port) {
+        AtomBlockId block = this->netlist.port_block(port);
+        this->netlist.set_block_param(block, "IOSTANDARD", io_standard);
+        VTR_LOG("XDC: Assigned IOSTANDARD %s = %s\n", this->netlist.block_name(block).c_str(), io_standard);
+    }
+
+    int _do_get_ports(const char* pin_name) {
+        AtomBlockId pin_block = this->netlist.find_block(pin_name);
+
+        if (pin_block == AtomBlockId::INVALID()) {
```

Comment:
Did you happen to bump in a case where the atom block ID could not be found?

Reply:
Of course. A typo in XDC. Giving a wrong port name will cause VPR to throw an exception and print an error saying that it can't find a block, followed by a list of available block names.

Reply:
Of course there's, but XDCs are provided by the user and we can't ensure they won't make a typo. We can however give them some meaningful information about the error they made.

I made VPR call `vpr_throw`/`VPR_THROW` in case of an error.